### PR TITLE
Github/Gitlab social auth

### PR DESCRIPTION
**Pull request by EDsCODE**
Addresses #404 

**Implemented:** When the `SOCIAL_AUTH_GITLAB_KEY` (or `SOCIAL_AUTH_GITHUB_KEY`)and `SOCIAL_AUTH_GITLAB_SECRET` (or `SOCIAL_AUTH_GITHUB_SECRET`) are provided as env variables wherever your instance is deployed, social auth with that platform will be available. The auth is implemented using python-social-auth as suggested. Currently, github and gitlab auth have been implemented.

**Note:** The redirect_uri that github or gitlab asks for should be in the form http://{domain_name}/complete/{gitlab or github}/

**What's left:** 
- style the buttons as they're quite barebones right now 
- tests?
- possible edge case where the social auth doesn't return email at all will break login

Screen Shot 2020-03-29 at 5 43 44 PM

### CODE REVIEW

**Comment by EDsCODE** [comment]
Screen Shot 2020-03-29 at 5 43 44 PM

**Comment by timgl** [comment]
This is awesome! Thank you so much for doing this. It looks good, will do a proper QA tomorrow.

I like the bare-bones styling of the buttons, it would be good if they aligned with the button above on the left and right

**Comment by timgl** [comment]
There's a couple of things missing:
- [ ] There's a bit of logic around [what happens after a user account is created](https://github.com/PostHog/posthog/blob/master/posthog/urls.py#L44) when a user signs up to a team that needs to be put into the [social auth pipeline](https://python-social-auth.readthedocs.io/en/latest/pipeline.html).
- [ ] I think we can safely [associate users by email](https://python-social-auth.readthedocs.io/en/latest/use_cases.html#associate-users-by-email) (if they signed up using their email before)

**Comment by EDsCODE** [comment]
> There's a couple of things missing:
> 
> * [ ]  There's a bit of logic around [what happens after a user account is created](https://github.com/PostHog/posthog/blob/master/posthog/urls.py#L44) when a user signs up to a team that needs to be put into the [social auth pipeline](https://python-social-auth.readthedocs.io/en/latest/pipeline.html).
> * [ ]  I think we can safely [associate users by email](https://python-social-auth.readthedocs.io/en/latest/use_cases.html#associate-users-by-email) (if they signed up using their email before)

Agreed! In fact, I was reviewing the implementation and realized I wasn't being thorough with the access points because I'm guessing these sign in options should be available on the initial admin sign up where a new team is created right? Going to add a custom pipeline to distinguish these signup/login paths.

**Comment by timgl** [comment]
I'm actually less concerned about the admin setup page, as I don't think anyone will have set up GitHub keys before setting up the master account, plus we still need to get the other fields (company name, full name etc) that we can't get easily from github/gitlab.

**Comment by EDsCODE** [comment]
hmm this will be a blocker for the standard login.html page too because if a new user trying to join a team ends up trying to auth from the default login page, social auth will complete but land them on a blank page as they're not associated with a team. There's two routes I can take with this PR:

1. (slightly lengthier) I can setup a unique flow where the admin setup and regular login page both have social auth capabilities. If a new user is joining (no team association), we can use a [partial pipeline ](https://python-social-auth.readthedocs.io/en/latest/pipeline.html)to redirect the user to a form where the user can enter a team token or create a new team.
2. (condensed) I can leave the social auth as it is now and just finish including the team addition in the pipeline as noted above. If a user who isn't associated with a team attempts to login from the normal page, they can be redirected to an error page noting that they're unable to login due to not belonging to a team.

I'd suggest 1 but 2 keeps this PR contained a bit better!

**Comment by timgl** [comment]
Good find! I much prefer 2 anyway as that feels right (who would know a team code of the top of their head?). We should make sure no user gets created, which I think you should be able to do by putting the function at the right point in the pipeline.

👍

File: posthog/settings.py [hunk 1]
```diff
@@ -117,6 +117,43 @@ def get_env(key):
...
+
+SOCIAL_AUTH_PIPELINE = (
+
+    'social_core.pipeline.social_auth.social_details',
+    'social_core.pipeline.social_auth.social_uid',
+    'social_core.pipeline.social_auth.auth_allowed',
+    'social_core.pipeline.social_auth.social_user',
+    'social_core.pipeline.user.get_username',
+    'social_core.pipeline.social_auth.associate_by_email',
+    'posthog.urls.social_create_user',
```
Comment (EDsCODE) [inline posthog/settings.py:139]: uprooted and used our own custom create user

File: posthog/urls.py [hunk 2]
```diff
@@ -92,6 +93,32 @@ def setup_admin(request):
         })
         return redirect('/')
 
+def social_create_user(strategy, details, backend, user=None, *args, **kwargs):
+    if user:
+        return {'is_new': False}
+
+    signup_token = strategy.session_get('signup_token')
+    if signup_token is None:
```
Comment (EDsCODE) [inline posthog/urls.py:101]: social auth will shortcircuit when a user is attempting to create an account without a team token. So, not socially authed users will exist without a team association

File: posthog/urls.py [hunk 3]
```diff
@@ -92,6 +93,32 @@ def setup_admin(request):
         })
         return redirect('/')
 
+def social_create_user(strategy, details, backend, user=None, *args, **kwargs):
```
Comment (EDsCODE) [inline posthog/urls.py:96]: extended the standard implementation of this function from the source. includes team addition now and so analytics for new users can be logged

**Comment by EDsCODE** [comment]
> I like the bare-bones styling of the buttons, it would be good if they aligned with the button above on the left and right

will try this right now!

File: posthog/urls.py (hunk 3 above)
Comment (EDsCODE) [reply posthog/urls.py:96]: when we need to implement social auth more thoroughly (through setup_admin, etc.) we can place logic here

Comment (timgl) [reply posthog/urls.py:96]: Great! Would be good to test this, either e2e (see `test_signup.py` for an example) or just this function.

File: posthog/urls.py (hunk 2 above)
Comment (timgl) [reply posthog/urls.py:101]: In terms of the error message here, I'm guessing you can get here if you're just trying to log in but don't actually have an account? In that case the error message is a bit confusing.

I think a user wouldn't come here if they signed up via the team signup page, so the user message could be something like "There's no account associated with that account" or something like that.

File: posthog/urls.py [hunk 4]
```diff
@@ -92,6 +93,32 @@ def setup_admin(request):
...
+        return HttpResponse(processed, status=401)
+
+    fields = dict((name, kwargs.get(name, details.get(name)))
+                   for name in backend.setting('USER_FIELDS', ['email']))
+    
+    if not fields:
+        return
+
+    user = strategy.create_user(**fields)
+    team = Team.objects.get(signup_token=signup_token)
```
Comment (timgl) [inline posthog/urls.py:112]: Would be good to catch `Team.DoesNotExist` and error into "We can't find the team associated with this signup token"

**Review by timgl — COMMENTED** [review]
QA'd, functionally works beautifully. Just two small things and then we're good to go 👍

**Review by timgl — APPROVED** [review]

**Pull request merged**